skimage/registration/_lddmm.py

```diff
@@ -0,0 +1,1247 @@
+    
+
+    def _generate_calibration_plots(self):
+        """
+        Plot the energies, maximum velocities, translation components, and linear components as functions of the number of iterations.
+        """
+
+        fig, axes = plt.subplots(2, 2, figsize=(6, 6))
+
+        # Plot matching, regularization, and total energies.
+        ax = axes[0, 0]
+        ax.plot(list(zip(self.matching_energies, self.regularization_energies, self.total_energies)))
+        ax.set_title('Energies')
+
+        # Plot the maximum velocity.
+        ax = axes[0, 1]
+        ax.plot(self.maximum_velocities)
+        ax.set_title('Maximum\nvelocity')
+
+        # Plot affine[:, :-1], the translation components.
+        translations = [affine[:-1, -1] for affine in self.affines]
+        ax = axes[1, 0]
+        ax.plot(translations)
+        ax.set_title('Translation\ncomponents')
+
+        # Plot self.affine[:-1, :-1], the linear transformation components.
+        linear_components = [affine[:-1, :-1].ravel() for affine in self.affines]
+        ax = axes[1, 1]
+        ax.plot(linear_components)
+        ax.set_title('Linear\ncomponents')
+
+    # End _Lddmm.
+
+r'''
+  _    _                          __                          _     _                       
+ | |  | |                        / _|                        | |   (_)                      
+ | |  | |  ___    ___   _ __    | |_   _   _   _ __     ___  | |_   _    ___    _ __    ___ 
+ | |  | | / __|  / _ \ | '__|   |  _| | | | | | '_ \   / __| | __| | |  / _ \  | '_ \  / __|
+ | |__| | \__ \ |  __/ | |      | |   | |_| | | | | | | (__  | |_  | | | (_) | | | | | \__ \
+  \____/  |___/  \___| |_|      |_|    \__,_| |_| |_|  \___|  \__| |_|  \___/  |_| |_| |___/
+                                                                                            
+'''
+
+def lddmm_register(
+    # Images.
+    template,
+    target,
+    # Image resolutions.
+    template_resolution=None,
+    target_resolution=None,
+    # Multiscale.
+    multiscales=None,
+    # Iterations.
+    num_iterations=None,
+    num_affine_only_iterations=None,
+    num_rigid_affine_iterations=None,
+    # Stepsizes.
+    affine_stepsize=None,
+    deformative_stepsize=None,
+    # Affine specifiers.
+    fixed_affine_scale=None,
+    # Velocity field specifiers.
+    sigma_regularization=None,
+    smooth_length=None,
+    num_timesteps=None,
+    # Contrast map specifiers.
+    contrast_order=None,
+    spatially_varying_contrast_map=None,
+    contrast_maxiter=None,
+    contrast_tolerance=None,
+    sigma_contrast=None,
+    # Artifact specifiers.
+    check_artifacts=None,
+    sigma_artifact=None,
+    # Smoothness vs. accuracy tradeoff.
+    sigma_matching=None,
+    # Initial values.
+    initial_affine=None,
+    initial_velocity_fields=None,
+    initial_contrast_coefficients=None,
+    # Diagnostic outputs.
+    calibrate=None,
+    track_progress_every_n=None,
+):
```

Comment:
As mentioned elsewhere, I think this signature is too complex for skimage. What can be simplified?

- can tracking be implicit using generators?
- how expensive is calibration output? Can we always return it?
- how critical is the split between rigid, affine, and total iterations? Can these be set to sensible defaults from a single `num_iter` argument?
- how are artifacts dealt with? Can we leave artifact handling to upstream functions?
- do we need to specify resolutions? Is it much more expensive to require users to upsample the lower-res volume or downsample the higher res one so that the resolutions match? Or even, require them to encode it in the initial affine?
- affine specifiers: these are constraints, right? Can we use the `vector_to_matrix=` from #3544 to encode these? That will nicely unify the interfaces. (Though note that that PR has been around for almost two years... 😬) See [here](https://github.com/scikit-image/scikit-image/pull/3544/files#diff-b794b66735c3941356f67f90f57c734cR104-R107), [here](https://github.com/scikit-image/scikit-image/pull/3544/files#diff-b794b66735c3941356f67f90f57c734cR61-R64), and [here](https://github.com/scikit-image/scikit-image/pull/3544/files#diff-a629bfe28e42b9b50c338c95ab5b5045R164-R173) for how that works.

Just some ideas to discuss...

Reply:
calibrate and track_progress_every_n have been removed as parameters. Tracking was removed completely, calibration output is now returned always because it is very cheap but images are no longer plotted, just the accumulated values that may be visualized (and are in the plot_lddmm.py example).

The separation between the num-iterations is to allow total flexibility. The actual numbers are problem-specific. rigid-affine should only be appropriate early in the registration to make any possible rotational adjustments and likely won't be used after. The number of affine-only iterations and deformative iterations should be readily determinable from visualizing the energies in the diagnostics output.